File: SourceCode/Python/Binary/ExpressionTree.py
#/* *** ODSATag: ExpressionTree *** */
# Base class for expression tree nodes
class VarBinNode:
    def isLeaf(self):  # All subclasses must implement
        raise NotImplementedError

# Leaf node
class VarLeafNode(VarBinNode):
    def __init__(self, val):
        self.operand = val
    def isLeaf(self):
        return True
    def value(self):
        return self.operand

# Internal node
class VarIntlNode(VarBinNode):
    def __init__(self, op, l, r):
        self.operator = op
        self.left = l
        self.right = r
    def isLeaf(self):
        return False
    def leftchild(self):
        return self.left
    def rightchild(self):
        return self.right
    def value(self):
        return self.operator

# Preorder traversal
def traverse(rt):
    if rt is None:
        return       # Nothing to visit
    if rt.isLeaf():  # Process leaf node
        VisitLeafNode(rt.value())
    else:            # Process internal node
        VisitInternalNode(rt.value())
        traverse(rt.leftchild())
        traverse(rt.rightchild())
#/* *** ODSAendTag: ExpressionTree *** */

File: SourceCode/Java_Generic/Binary/ExpressionTree.java
/* *** ODSATag: ExpressionTree *** */
/* *** ODSATag: ExpressionTree1 *** */
// Base class for expression tree nodes
public interface VarBinNode {
    public boolean isLeaf(); // All subclasses must implement
}

/** Leaf node */
public class VarLeafNode implements VarBinNode {
    private String operand;        // Operand value

    VarLeafNode(String val) {
        operand = val;
    }

    public boolean isLeaf() { return true; }
    public String  value()  { return operand; }
}
/* *** ODSAendTag: ExpressionTree1 *** */

/* *** ODSATag: ExpressionTree2 *** */
// Internal node
public class VarIntlNode implements VarBinNode {
    private VarBinNode left;       // Left child
    private VarBinNode right;      // Right child
    private Character  operator;   // Operator value

    VarIntlNode(Character op, VarBinNode l, VarBinNode r) {
        operator = op; left = l; right = r;
    }

    public boolean    isLeaf()     { return false; }
    public VarBinNode leftchild()  { return left; }
    public VarBinNode rightchild() { return right; }
    public Character  value()      { return operator; }
}
/* *** ODSAendTag: ExpressionTree2 *** */

// Preorder traversal
/* *** ODSATag: pointer based preorder *** */
public static void traverse(VarBinNode node) {
    if (node == null) return;  // Nothing to visit
    if (node.isLeaf()) {
        // Process leaf node
        Visit.VisitLeafNode(((VarLeafNode)node).value());
    } else {
        // Process internal node
        Visit.VisitInternalNode(((VarIntlNode)node).value());
        traverse(((VarIntlNode)node).leftchild());
        traverse(((VarIntlNode)node).rightchild());
    }
}
/* *** ODSAendTag: pointer based preorder *** */
/* *** ODSAendTag: ExpressionTree *** */
